pkg/k8s/watchers/cilium_node.go

```diff
@@ -5,167 +5,167 @@ package watchers
-						}
-					}
-				},
-				DeleteFunc: func(obj interface{}) {
-					var equal bool
-					defer func() { k.K8sEventReceived(apiGroup, metricCiliumNode, resources.MetricDelete, true, equal) }()
-					ciliumNode := k8s.CastInformerEvent[cilium_v2.CiliumNode](obj)
-					if ciliumNode == nil {
-						return
-					}
-					n := nodeTypes.ParseCiliumNode(ciliumNode)
-					k.nodeDiscoverManager.NodeDeleted(n)
-				},
-			},
+		var synced atomic.Bool
+		stop := make(chan struct{})
+
+		k.blockWaitGroupToSyncResources(
+			stop,
 			nil,
+			func() bool { return synced.Load() },
+			apiGroup,
 		)
-		isConnected := make(chan struct{})
-		// once isConnected is closed, it will stop waiting on caches to be
-		// synchronized.
-		k.blockWaitGroupToSyncResources(isConnected, swgNodes, ciliumNodeInformer.HasSynced, apiGroup)
-
-		k.ciliumNodeStoreMU.Lock()
-		k.ciliumNodeStore = ciliumNodeStore
-		k.ciliumNodeStoreMU.Unlock()
-
-		once.Do(func() {
-			// Signalize that we have put node controller in the wait group
-			// to sync resources.
-			asyncControllers.Done()
-		})
 		k.k8sAPIGroups.AddAPI(apiGroup)
-		go ciliumNodeInformer.Run(isConnected)
 
-		<-kvstore.Connected()
-		log.Info("Connected to key-value store, stopping CiliumNode watcher")
+		// Signalize that we have put node controller in the wait group to sync resources.
+		once.Do(asyncControllers.Done)
+
+		// derive another context to signal Events() and Store() in case of kvstore connection
+		subCtx, cancel := context.WithCancel(ctx)
+
+		var wg sync.WaitGroup
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer close(stop)
+
+			events := k.resources.CiliumNode.Events(subCtx)
+			cache := make(map[resource.Key]*cilium_v2.CiliumNode)
+			for event := range events {
+				var err error
+				switch event.Kind {
+				case resource.Sync:
+					synced.Store(true)
+				case resource.Upsert:
+					var needUpdate bool
+					oldObj, ok := cache[event.Key]
+					if !ok {
+						needUpdate = k.onCiliumNodeInsert(event.Object)
+					} else {
+						needUpdate = k.onCiliumNodeUpdate(oldObj, event.Object)
+					}
+					if needUpdate {
+						cache[event.Key] = event.Object.DeepCopy()
+					}
+				case resource.Delete:
+					k.onCiliumNodeDelete(event.Object)
+					delete(cache, event.Key)
+				}
+				event.Done(err)
+			}
+		}()
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
 
-		// Set the ciliumNodeStore as nil so that any attempts of getting the
-		// CiliumNode are performed with a request sent to kube-apiserver
-		// directly instead of relying on an outdated version of the CiliumNode
-		// in this cache.
-		k.ciliumNodeStoreMU.Lock()
-		k.ciliumNodeStore = nil
-		k.ciliumNodeStoreMU.Unlock()
+			store, err := k.resources.CiliumNode.Store(subCtx)
+			if err != nil {
+				if !errors.Is(err, context.Canceled) {
+					log.WithError(err).Warning("unable to retrieve CiliumNode local store, going to query kube-apiserver directly")
```

Comment:
Bit confused about this, where does the directly querying apiserver take place in this scenario?

Reply:
At line 157. This might happen if `Store` fails or:

- at startup, if `GetCiliumNode` is called before the store is synchronized
- if kvstore is connected and the informer is stopped